### DISCUSSION TITLE
asyncio.web.Application - await on incoming requests, instead of handler callbacks

### DISCUSSION THREAD

**Discussion by haxtibal** [aio-libs/aiohttp#6535] [Q&A] [answered]
Hi folks,

`asyncio.web.Application` wants us to register callbacks to process incoming HTTP requests. However I have a use case where I'd prefer to treat incoming (HTTP POST) requests as events, and would like to have an async generator interface to await on such events in a `async for` loop, like
```python
async for posted_event in http_io_channel.event_stream():
    the_core.push_event(posted_event)
```

We can of course use `asyncio.Future`, `asyncio.Queue`, `asyncio.Event`, ... to convert `aiohttp`'s callback API into a generator API on our own. A simple future based solution may look like

```python
...

event_stream_futures = []

async def event_stream(self):
   """Custom generator interface."""
    while True:
        future = asyncio.Future()
        event_stream_futures.append(future)
        await future
        yield future.result()

async def post_handler(request):
   """Callback as wanted by aiohttp."""
    body = await request.text()
    prev_futures = event_stream_futures
    event_stream_futures  = []
    for future in prev_futures:
        future.set_result(convert_checkresult_body_to_event(body))
    return aiohttp.web.Response()
```

Is this the way to go? Or is awaiting on upcoming requests already built into `aiohttp` and I just haven't discovered it?

**Comment by asvetlov** [comment, marked as answer]
There is no embedded queue.
A custom queue-based approach should work, sure.
